Triggers.md

```diff
@@ -0,0 +1,234 @@
+
+### New Commands
+1. `CLIENT TRIGGERS [ON|OFF]` new subcommand. Sometimes it will be required to mute trigger execution in some context (say administrative fix to the dataset) 
+    In order to support it the specific client will be able to mute triggers execution.
+    Response will always be "O.K".  
+2. The `FUNCTION LIST` command output will now include a new part named "triggers".
+Much like functions the triggers part will describe a per-trigger information:
+
+```
+"library_name"
+  "mylib"
+  "engine"
+  "LUA"
+  "functions"
+         1) "name"
+         2) "myfunc"
+         3) "description"
+         4) <some description>
+         5) "flags"
+         6) <relevant flags>
+   "triggers"
+          1) "name"
+          2) "trigger_name"
+          3) "event"
+          4) "__keyspace@0__:*"
+          7) "calls"
+          8) "<the total number of times the trigger was called>"
+          9) "errors"
+         10) "<the total number of errors issued during this trigger run>"
+         11) "total_duration"
+         12) <the aggregated total duration time this trigger run>
+``` 
+
+### Benchmarking (Optional)
+
+TBD 
+
+### Testing 
+TBD
+
+### Observability 
+
+#### general statistics
+The "Stats" info section will be added with the following statistics:
+**total_trigger_calls -** the number of general trigger calls. this will not be a teardown list of calls per trigger function (this can be taken from the function list command as will be explained later)
+**total_trigger_errors -** The total number of errors during trigger execution. (an error is accounted for each time the afterErrorReply is called) 
+**total_trigger_duration -** The total time spent running trigger related code.
+
+### Per trigger statistics
+The same global statistics will be available on a per-trigger resolution via the `FUNCTION LIST` command:
+"calls" -  the total number of times the trigger has called
+"errors" - the total number of errors issued during this trigger run
+"total_duration" - the aggregated total duration time this trigger run
+
+*** Open question ***
+should `CONFIG RESETSTAT` also reset the triggers statistics, or should that only be done when functions lib is reloaded?
+
+### Debug mechanism 
+
+While Valkey functions and scripts are executed as part of the FCALL/EVAL calls, it is thus possible to report the errors back to the calling client.
+Triggers, however, are silently executing which makes it hard for the user to understand why the trigger did not work and what errors occurred during execution of the trigger.
+The suggestion here is to enable reporting error msg on a predefined pub/sub channel:
+1. All error msgs will be intercepted as deferred errors (much like modules do). 
+2. After trigger completes run, in case errors were issues, it will update the stats with the num,ber of errors.
+3. The trigger will report the errors to a dedicated channel in the following structure:
+   `__triggers@errors__:<trigger-name>`
```

Comment:
Since the trigger execution is hidden from the user we cannot just return it as an error. 

For the case when the trigger is issued based on internal events (eg evictions) there is no request context to issue the error to.
For case were the event was issued as part of the complex command flow (eg script or transaction),
we have 2 options to consider trigger execution:
1. without breaking the action atomicity (in which case the trigger will be executed AFTER the command was executed) in which case the response to the command was already sent or written to the COB, so not much point or ability to return it to the client.
2. we can execute the trigger as soon as the event is reported, practically breaking the atomicity and in this case I agree there is a problem we need to consider how to handle trigger failure, since it potentially already changed the data.